<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Sean Harrington &amp; Vikram Chhatre" />
  <meta name="dcterms.date" content="2021-10-29" />
  <title>iPyRAD assembly of data for Crotalus ruber</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <link rel="stylesheet" href="avenir-white.css" />
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">iPyRAD assembly of data for <em>Crotalus ruber</em></h1>
<p class="author">Sean Harrington &amp; Vikram Chhatre</p>
<p class="date">October 29, 2021</p>
</header>
<h2 id="table-of-contents">Table of Contents</h2>
<ul>
<li><p><a href="#introduction">1. Introduction</a></p></li>
<li><p><a href="#files-and-basic-setup">2. Files and Basic Setup</a></p></li>
<li><p><a href="#installing-ipyrad">3. Installing iPyRad</a></p></li>
<li><p><a href="#running-ipyrad">4. Running iPyRad</a></p></li>
<li><p><a href="#branching-an-assembly">5. Branching An Assembly</a></p></li>
<li><p><a href="#examining-the-output">6. Examining the output</a></p>
<ul>
<li><a href="#do-another-round-of-branching">6.1 Do Another Round of Branching</a></li>
</ul></li>
</ul>
<br>
<center>
<img src="Crotalus_ruber_42613167.jpg" width=800 />
</center>
<p><br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br></p>
<h2 id="introduction">1. Introduction</h2>
<p><a href="https://ipyrad.readthedocs.io/en/master/">iPyRad</a> is a flexible python-based pipeline for taking various types of restriction-site associated data, processing them, and generated aligned datasets.</p>
<p>iPyRad is capable of generating datasets either by mapping your raw reads to a reference genome or using a de novo assembly method that does not require a reference. We will use the reference-based method today.</p>
<p>If you need help with iPyRad outside of this workshop for specific issues, you can always post <a href="https://gitter.im/dereneaton/ipyrad">here</a>. Isaac is <em>very</em> responsive to queries.</p>
<p><br><br><br></p>
<h2 id="files-and-basic-setup">2. Files and Basic Setup</h2>
<p>Before we get started, everyone will need the following files:</p>
<ul>
<li><code>all_ruber.fastq</code></li>
<li><code>barcodes_samples.txt</code></li>
<li><code>GCA_003400415.2_UTA_CroVir_3.0_genomic.fna</code></li>
<li><code>names_ruber_all.txt</code></li>
</ul>
<p>These are all contained on Teton at <code>/project/inbre-train/2021_popgen_wkshp/data</code></p>
<p>To copy these over to your directory, use</p>
<pre><code>cp -r /project/inbre-train/2021_popgen_wkshp/data &lt;where_you_want_these_files&gt;</code></pre>
<p>Let’s then also create a new directory for the output of our iPyRad assembly. This will make several directories, and it’s nice to have them self-contained</p>
<pre><code>mkdir ipyrad_out
cd ipyrad_out</code></pre>
<p>We’ll run this interactively (as long as everyone can get sessions allocated).</p>
<pre><code>salloc --account=&lt;YOUR_ACCOUNT&gt; -t 0-2:00:00 --mem=10G --nodes=1 --ntasks-per-node=6</code></pre>
<p>If you are unable to quickly get a session allocated with 6 cores, try <code>--ntasks-per-node=4</code> instead</p>
<p><br><br><br></p>
<h2 id="installing-ipyrad">3. Installing iPyRad</h2>
<p>We will use conda to install iPyRad and all of its dependencies. To do this, we first have to load up the Teton modules necessary for miniconda</p>
<pre><code>module load swset/2018.05 gcc/7.3.0 miniconda3/4.9.2</code></pre>
<p>Create a new conda environemtn called ipyrad, activate the environment, and install iPyRad</p>
<pre><code>conda create -n ipyrad  # create the environment: we only need to do this once
conda activate ipyrad  ### We will need to run this command every time we want to use ipyrad
conda install ipyrad -c conda-forge -c bioconda  ## We only need to run this once</code></pre>
<p><br><br><br></p>
<h2 id="running-ipyrad">4. Running iPyRad</h2>
<p>First, we need to generate a params file that contains the parameters we need to specify for ipyrad:</p>
<pre><code>ipyrad -n ruber_ref   </code></pre>
<p>This will create a params file with the defaults that ipyrad uses, we can modify these as we need . Whatever comes after the -n is what the assembly will be named</p>
<p>Let’s go look at and edit that</p>
<pre><code>nano params-ruber_ref.txt  #if you prefer a different text editor to nano, use that instead</code></pre>
<p>We’ll change a few of these: - <code>[0]</code>: change assembly name to <code>ruber_ref</code></p>
<ul>
<li><p><code>[2]</code>: this needs to reflect the path to the <code>all_ruber.fastq</code> file wherever it lives for you</p></li>
<li><p><code>[3]</code>: this needs to be the path to <code>barcodes_samples.txt</code></p></li>
<li><p><code>[5]</code>: change to <code>reference</code></p></li>
<li><p><code>[6]</code>: change to the path to <code>GCA_003400415.2_UTA_CroVir_3.0_genomic.fna</code></p></li>
<li><p><code>[7]</code>: dataype should be <code>ddrad</code></p></li>
<li><p><code>[8]</code>: restriction overhang is: <code>TGCAGG, GATC</code> these are the overhangs created by the restriction enzymes for ddRAD that was used for these data</p></li>
<li><p><code>[27]</code>: change to <code>*</code>, this will generate all output formats that ipyrad is currently capable of</p></li>
</ul>
<p>The rest of these are at generally reasonable values, although depending on your data, you may want to modify some of these. The parameters are all well <a href="https://ipyrad.readthedocs.io/en/master/6-params.html"><strong>documented here</strong></a></p>
<p>Save the file (<code>ctrl + x</code>)</p>
<p>Run this for <code>steps 1-5</code></p>
<pre><code>ipyrad -p params-ruber_ref.txt -s 12345 -c 6 # If you had to use 4 cores instead of 6 for salloc, then change this to -c 4</code></pre>
<p>We’re running this interactively, and it should take around 20 minutes. If you had trouble with getting an interactive <code>salloc</code> session running, you could also set this up as a job instead.</p>
<p>Instructions for how one would do this:</p>
<p>create and open a new file to run a slurm job (e.g., <code>ruber_ref.slurm</code>). It should contain the following, with the starred sections for your email and your account replaced with your information:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="co">#!/bin/bash</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a><span class="co">#SBATCH --job-name ruber_denovo</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a><span class="co">#SBATCH -A *****YOUR_ACCOUNT*******</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a><span class="co">#SBATCH -p teton</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a><span class="co">#SBATCH -t 2-00:00</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a><span class="co">#SBATCH --cpus-per-task=8</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a><span class="co">#SBATCH --mem=10G</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="co">#SBATCH --mail-type=ALL</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a><span class="co">#SBATCH --mail-user=*****YOUR_EMAIL*********</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="co">#SBATCH -e ruber_%A_%a.err</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a><span class="co">#SBATCH -o ruber_%A_%a.out</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a><span class="ex">module</span> load swset/2018.05 gcc/7.3.0 miniconda3/4.9.2</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true"></a><span class="co">## run ipyrad</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true"></a><span class="ex">conda</span> activate ipyrad</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true"></a><span class="ex">ipyrad</span> -p params-ruber_denovo.txt -s 12345 -c 8</span></code></pre></div>
<p>Save that file, then submit the job:</p>
<pre><code>sbatch ruber_denovo.slurm</code></pre>
<p>While our job is running, we can take a look at what’s happening in each of the iPyRad steps. These are <a href="https://ipyrad.readthedocs.io/en/master/7-outline.html"><strong>documented here</strong></a>.</p>
<p><br><br><br></p>
<h2 id="branching-an-assembly">5. Branching an assembly</h2>
<p>The FASTQ that we started with includes most individuals of the red diamond rattlesnake, <em>Crotalus ruber</em> but also a few outgroup samples. For the types of analyses we’ll be doing downstream, we want to include only <em>C. ruber</em> samples.</p>
<p>iPyRad includes functionality to make new “branches” of the assembly using different parameters and/or including/excluding different individuals, and we’ll take advantage of that functionality here.</p>
<p>Make a branch of the reference assembly</p>
<pre><code>ipyrad -p params-ruber_ref.txt -b ruber_only_ref names_ruber_all.txt</code></pre>
<p>This will use our old assembly and params file to generate a new branch, with params file “params-ruber_only_ref.txt” that includes only samples in the “names_ruber_all.txt” file.</p>
<p>We need to further edit this file so that “min_sample per locus” is set to 26 - this is about 75% of individuals. This is param [21] - should be at 4 by default.</p>
<pre><code>nano params-ruber_only_ref.txt</code></pre>
<p>Once that change has been made, run the final 2 steps in ipyrad:</p>
<pre><code>ipyrad -p params-ruber_only_ref.txt -s 67 -c 4 </code></pre>
<p><br><br><br></p>
<h2 id="examining-the-output">6. Examining the output</h2>
<pre><code>cd ruber_only_ref_outfiles
less -S ruber_only_ref_stats.txt</code></pre>
<p>There should be 2825 loci recovered in the assembly (or something very similar). If we scroll down a bit, we can see that <code>SD_Field_0506</code> has almost no loci shared with other samples, and <code>SD_Field_1453</code> has only about half as many loci as most samples. We’ll want to remove these samples before moving on. Note that <code>SD_Field_0506</code> is an obviously failed sample, but for <code>SD_Field_1453</code>, you would likely want to try out some preliminary downstream analyses with and without this sample–I’ve already played with these data and decided it’s best to remove it.</p>
<h3 id="do-another-round-of-branching">6.1 Do Another Round of Branching</h3>
<p>Start by making a new names file to exclude <code>SD_Field_0506</code> and <code>SD_Field_1453</code></p>
<pre><code>cd ..
cp names_ruber_all.txt names_ruber_reduced.txt
nano names_ruber_reduced.txt  # delete the lines with the poorly assembled individuals</code></pre>
<p>Do the branching:</p>
<pre><code>ipyrad -p params-ruber_only_ref.txt -b ruber_reduced_ref names_ruber_reduced.txt</code></pre>
<p>Run step 7 on this new branch:</p>
<pre><code>ipyrad -p params-ruber_reduced_ref.txt -s 7 -c 4</code></pre>
<p>Let’s go look at the stats files for each new assembly</p>
<pre><code>less -S ruber_reduced_ref_outfiles/ruber_reduced_ref_stats.txt</code></pre>
<p>The assembly is down to 2783 loci, a slight decrease, but we now pretty good coverage across individuals.</p>
<p><br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br> <br><br><br><br><br><br></p>
</body>
</html>
